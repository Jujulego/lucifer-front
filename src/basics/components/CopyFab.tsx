import React, { MouseEvent, useState, useEffect } from 'react';
import copy from 'copy-to-clipboard';
import clsx from 'clsx';
import { pick } from 'lodash';

import {
  Fade,
  Fab, FabProps, FabClassKey
} from '@material-ui/core';
import { makeStyles, Theme } from '@material-ui/core/styles';
import CheckIcon from '@material-ui/icons/Check';
import FileCopyIcon from '@material-ui/icons/FileCopy';

import { StyledProps } from 'utils/style';

// Types
export type CopyFabClassKey = FabClassKey | 'succeed';
export type CopyFabProps = Omit<FabProps, 'classes' | 'children'> & StyledProps<CopyFabClassKey> & {
  text: string;
  format?: string;

  onCopied?: () => void;
}

// Styles
const useStyles = makeStyles(({ palette }: Theme) => ({
  label: {
    display: 'grid',
    justifyItems: 'center',

    '& > *': {
      gridArea: '1 / 1 / 2 / 2',
    }
  },
  succeed: {
    backgroundColor: palette.success.main,

    '&:hover': {
      backgroundColor: palette.success.dark,
    }
  }
}));

// Component
const CopyFab = (props: CopyFabProps) => {
  // Props
  const {
    text, classes,
    format = "text/plain",
    onCopied, onClick,
    ...btn
  } = props;

  // State
  const [succeed, setSucceed] = useState(false);

  // Effects
  useEffect(() => {
    if (succeed) {
      const id = setTimeout(() => setSucceed(false), 2000);
      return () => clearTimeout(id);
    }
  }, [succeed, setSucceed]);

  // Handler
  const handleCopy = (event: MouseEvent<HTMLButtonElement>) => {
    if (onClick) onClick(event);

    // Copy
    const success = copy(text, { format });
    if (success) {
      setSucceed(true);
      if (onCopied) onCopied();
    }
  };

  // Render
  const styles = useStyles({ classes: pick(classes, ['succeed', 'label']) });

  return (
    <Fab
      {...btn} onClick={handleCopy}
      classes={{
        ...classes,
        root: clsx(classes?.root, { [styles.succeed]: succeed }),
        label: styles.label,
      }}
    >
      <Fade in={succeed} timeout={250}>
        <CheckIcon />
      </Fade>
      <Fade in={!succeed} timeout={250}>
        <FileCopyIcon fontSize="small" />
      </Fade>
    </Fab>
  );
};

export default CopyFab;
