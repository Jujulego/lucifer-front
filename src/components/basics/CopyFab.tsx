import React, { MouseEvent } from 'react';
import copy from 'copy-to-clipboard';

import { Fab, FabProps } from '@material-ui/core';
import FileCopyIcon from '@material-ui/icons/FileCopy';

// Types
export interface CopyFabProps extends FabProps {
  text: string;
  format?: string;

  onCopied?: () => void;
}

// Component
const CopyFab = (props: CopyFabProps) => {
  // Props
  const {
    text, format = "text/plain",
    onCopied, onClick,
    ...btn
  } = props;

  // Handler
  const handleCopy = (event: MouseEvent<HTMLButtonElement>) => {
    if (onClick) onClick(event);

    // Copy
    const success = copy(text, { format });
    if (success && onCopied) onCopied();
  };

  // Render
  return (
    <Fab {...btn} onClick={handleCopy}>
      <FileCopyIcon fontSize="small" />
    </Fab>
  );
};

export default CopyFab;