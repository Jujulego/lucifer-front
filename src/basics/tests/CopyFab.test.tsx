import React from 'react';
import copy from 'copy-to-clipboard';
import { act } from 'react-dom/test-utils';
import { createShallow, createMount } from '@material-ui/core/test-utils';

import CopyFab from '../components/CopyFab';

// Mocks
jest.mock('copy-to-clipboard');

// Setup
let mount: ReturnType<typeof createMount>;
let shallow: ReturnType<typeof createShallow>;

beforeAll(() => {
  jest.useFakeTimers();

  mount = createMount();
  shallow = createShallow();
});

afterAll(() => {
  jest.restoreAllMocks();

  mount.cleanUp();
});

beforeEach(() => {
  (copy as jest.Mock).mockReset();
});

// Tests
it('should render correctly', () => {
  // Render
  const wrapper = shallow(
    <CopyFab text="test" />
  );

  // Check elements
  expect(wrapper).toMatchSnapshot();
});

it('should copy on click', () => {
  const spy = jest.fn();
  (copy as jest.Mock).mockImplementation(() => true);

  // Render
  const wrapper = mount(
    <CopyFab
      text='test' format='text/plain'
      onCopied={spy}
    />
  );

  // Get button
  const button = wrapper.find('button');
  expect(button).toHaveLength(1);

  // Test event
  act(() => {
    button.simulate('click');
    jest.runAllTimers();
  });

  expect(copy).toHaveBeenCalledTimes(1);
  expect(copy).toHaveBeenCalledWith('test', { format: 'text/plain' });

  expect(spy).toHaveBeenCalledTimes(1);
});
