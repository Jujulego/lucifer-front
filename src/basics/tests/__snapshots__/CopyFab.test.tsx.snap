// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<WithStyles(ForwardRef(Fab))
  classes={
    Object {
      "label": "makeStyles-label-1",
      "root": "",
    }
  }
  onClick={[Function]}
>
  <ForwardRef(Fade)
    in={false}
    timeout={250}
  >
    <CheckIcon />
  </ForwardRef(Fade)>
  <ForwardRef(Fade)
    in={true}
    timeout={250}
  >
    <FileCopyIcon
      fontSize="small"
    />
  </ForwardRef(Fade)>
</WithStyles(ForwardRef(Fab))>
`;
